runtime/doc/map.txt

```diff
@@ -1431,6 +1431,211 @@ Possible values are (second column is the short name used in listing):
+steps, in order, to implement the Nvim command preview "protocol".
+    1. Make necessary buffer changes and highlighting for the preview.
+    2. If 'inccommand' equals "split" and Nvim is in command preview mode,
+       open preview window using |vim.cmd_preview.open_win()| and put the
+       necessary text and highlighting on it. Otherwise, close preview window
+       if it's open using |vim.cmd_preview.close_win()|.
+    3. Trigger a |:redraw| of the window.
+    4. Undo all changes made to the buffer in step 1, without keeping them
+       in the undo history (see |:undo|).
+    5. Clear all highlights set in step 1.
+    6. If preview window was opened in step 2, switch back to main window to
+       ensure that if the command is executed, it gets executed on the main
+       window and not the preview window.
+
+Here is an example written in |Lua| that uses the preview flag to preview a
+command that trims trailing whitespaces from lines while following the
+aforementioned steps: >
+
+	-- Namespace for command preview highlights
+	local namespace =
+		vim.api.nvim_create_namespace('TrimTrailingWhitespaceNS')
+	local preview_win = 0
+
+	-- Function that determines whether command preview is triggered
+	local function PreviewFunc(cmd)
+	    -- Preview when first argument equals 'enable'
+	    local args = vim.fn.split(cmd, ' ')
+	    return #args > 1 and args[2] == 'enable'
+	end
+
+	local function TrimTrailingWhitespace(opts)
+	    local line1 = opts.line1
+	    local line2 = opts.line2
+	    local curwin = vim.api.nvim_get_current_win()
+	    local curbuf = vim.api.nvim_get_current_buf()
+	    local lines = vim.api.nvim_buf_get_lines(curbuf, line1 - 1,
+						     line2, 0)
+	    local new_lines = {}
+
+	    -- Check the first two letters of the mode string rather than the
+	    -- whole mode string since more specific modes may be added
+	    -- at any time
+	    local cmdpreview = vim.api.nvim_get_mode().mode:sub(1, 2) == "cp"
+	    local incsplit = vim.opt.inccommand:get() == 'split'
+
+	    -- If in command preview mode, temporarily unset 'cursorcolumn',
+	    -- 'cursorline' and 'hlsearch' to not mess up highlights
+	    local save_cursorcolumn = vim.o.cursorcolumn
+	    local save_cursorline = vim.o.cursorline
+	    local save_hlsearch = vim.o.hlsearch
+	    vim.o.cursorcolumn = false
+	    vim.o.cursorline = false
+	    vim.o.hlsearch = false
+
+	    -- If in command preview mode and inccommand=split,
+	    -- open the preview window if it's not already opened
+	    -- Otherwise close preview window
+	    if cmdpreview and incsplit then
+		if not (preview_win ~= 0
+			and vim.fn.win_gettype(preview_win) ~= 'unknown')
+		then
+		    preview_win = vim.cmd_preview.open_win()
+		end
+	    else
+		preview_win = 0
+		vim.cmd_preview.close_win()
+	    end
+
+	    local preview_buf = vim.api.nvim_win_get_buf(preview_win)
+	    local preview_buf_line = 0
+
+	    for i = 0, #lines - 1 do
```

Comment:
Any reason you are not just using `for i, line in ipairs(lines)`?

Reply:
Ah nope. This function was translated from Vimscript to Lua since the original one was written in Vimscript, so I ended up translating that literally. My bad